MM-54238 Add WebSocket broadcast hook and don't broadcast when other users were mentioned

hmhealey opened the pull request:
#### Summary
The main motivation for this is that we don't want to broadcast in WebSocket events who else was notified by a post. This data is contained as an array of user IDs in the `mentions` and `followers` properties of a `posted` WebSocket event. To avoid changing the structure of the events and causing compatibility issues with mobile, this makes it so that both of those fields will either be empty or an array containing just the current user ID.

To be efficient about how we do this, I copied what Agniva was doing in #23812 by introducing hooks to process WS events right before they're sent to a client. This allows us to publish a single event from the server and have its data transformed for individual users which is important when an event like posting in a large channel may trigger thousands of WS events. That PR was reverted due to issues serializing function calls in a multi-server environments, so this passes a string ID representing the hook and a map of serializable arguments which should work with multiple servers.

#### Ticket Link
MM-54238

#### Release Note
```release-note
Added new way to modify WebSocket messages sent to individual connections
```

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

hmhealey commented on server/channels/app/notification.go:
```diff
@@ -518,12 +518,11 @@ func (a *App) SendNotifications(c request.CTX, post *model.Post, team *model.Tea
-	if len(notificationsForCRT.Desktop) != 0 {
-		message.Add("followers", model.ArrayToJSON(notificationsForCRT.Desktop))
+	if len(mentionedUsersList) > 0 || len(notificationsForCRT.Desktop) > 0 {
+		message.GetBroadcast().AddHook(AddMentionsAndFollowers, map[string]any{
```
Note that the event doesn't actually start with the `mentions` or `followers` fields, so even if the hook doesn't run like in a mixed version cluster which doesn't understand broadcast hooks, it'll still send the safer version of the event to clients

hmhealey commented on server/channels/app/platform/web_broadcast_hook.go:
```diff
@@ -0,0 +1,52 @@
+import "github.com/mattermost/mattermost/server/public/model"
+
+type BroadcastHook struct {
+	HasChanges func(msg *model.WebSocketEvent, webConn *WebConn, args map[string]any) bool
+	Process    func(msg *model.WebSocketEvent, webConn *WebConn, args map[string]any) *model.WebSocketEvent
```
I broke up the hook into two methods to avoid allocating unnecessary `model.WebSocketEvent`s every time an event is published with a hook attached. This seems important for cases like the `posted` event where a post in a busy channel may cause thousands of events to be broadcast when only a few users are mentioned.

I could've made it so that `BroadcastHook.Process` is required to return a new object if it makes changes, but I was wanting to make it so that hook implementors won't have to worry about that. It's only a shallow copy, so they do still have to use `AddWithCopy` and not mutate the inner fields, but this feel slightly nicer at least.

I'm happy to change this though if anyone has any strong opinions

hmhealey commented on server/channels/app/platform/web_hub.go:
```diff
@@ -497,6 +505,8 @@ func (h *Hub) Start() {
 						return
 					}
 					if webConn.ShouldSendEvent(msg) {
+						msg := h.runBroadcastHooks(msg, webConn)
```
Broadcast hooks are the last thing that happens before we send the event from the hub to an individual connection so that they'll be called for each event when something is broadcast to a whole channel or team

hmhealey commented on server/channels/app/websocket_broadcast_hooks.go:
```diff
@@ -0,0 +1,59 @@
+				return false
+			}
+
+			mentions, _ := args["mentions"].(model.StringArray)
```
I was trying to make sure that this code was type safe, but because I couldn't figure out how to make the type of `args` narrower than `map[string]any`, even with generics, I went with just adding the second return value to suppress panics if the argument is missing

hmhealey replied:
I can change it to be more idiomatic then. I would like if I could make this type check happen as a compile time check to avoid needing any of this at all, but I'm still not sure how to accomplish that even with generics.

If I do that, should I change this method to return an error if that happens, or should I just log an error and move on when it occurs?

hmhealey commented on server/channels/app/server.go:
```diff
@@ -275,6 +275,8 @@ func NewServer(options ...Option) (*Server, error) {
 	// Step 4: Start platform
 	s.platform.Start()
 
+	s.RegisterBroadcastHooks()
```
I'm not happy with having to do this here, but`s.platform.Start()` also starts the websockets and I couldn't find a  way to pass the `broadcastHooks` map into it that felt proper

hmhealey replied:
If I do that, I'd also be moving the hooks implementation into the platform package which I was trying to avoid since it feels like more of an app concern. Is that okay?

agnivade replied:
Good point. We can simply pass the hooks map directly from `platform.Start`, and pass them down to `hubStart`.

agnivade commented:
Give me some time on this Harrison.

agnivade commented on server/channels/app/notification_test.go:
```diff
@@ -212,6 +216,268 @@ func TestSendNotifications(t *testing.T) {
+	th.LinkUserToTeam(sender, th.BasicTeam)
+	member := th.AddUserToChannel(sender, th.BasicChannel)
+
+	// Make the sender a channel_admin because that's needed for group mentions
```
This seems wrong. Should we investigate this further with a ticket?

hmhealey replied:
I raised a discussion for it [here](https://community.mattermost.com/core/pl/rtkkr8fsb7nebpd6ub9go16ydh). It turns out that the default permissions are that only channel admins can use group mentions

isacikgoz replied:
Commented on the thread, I think we should investigate it.. At least we can add this permission/role only when it's necessary in the `L:295`?

hmhealey replied:
Will do

agnivade commented on server/channels/app/notification_test.go:
```diff
@@ -212,6 +216,268 @@ func TestSendNotifications(t *testing.T) {
+	})
+}
+
+func connectFakeWebSocket(t *testing.T, th *TestHelper, userID string, connectionID string) (chan *model.WebSocketEvent, func()) {
```
Any reason we cannot copy over `th.CreateWebSocketClient` from api layer to this? See `TestPostReminder` for example. This code is re-doing lot of the initialization code that happens inside the web_conn.go file. Any changes to that, will need to be copied over to this again. I'd rather have something high level to interact with the websockets.

hmhealey replied:
I can see if I can copy that over. I don't know if we can access the API endpoint here, but it'd be nice if we can use the same interface as that method at least

hmhealey replied:
I spent a couple hours trying to make this work, and I couldn't manage to make this code use a `model.WebSocketClient` since it makes a lot of assumptions about there being an HTTP API to connect the WebSocket. This function also replicates the logic from `api4/websocket.go` which sets connects the `WebConn` and isn't part of the `app` package, so we'd either need to move that to the app layer or duplicate it here. I've spent enough time on this that I think I need to move on from this one

hmhealey commented on server/public/model/websocket_message.go:
```diff
@@ -104,6 +104,11 @@ type WebsocketBroadcast struct {
+
+	// BroadcastHooks is a slice of hooks IDs used to process events before sending them on individual connections. The
+	// IDs should be understood by the WebSocket code.
+	BroadcastHooks    []string         `json:"-"`
+	BroadcastHookArgs []map[string]any `json:"-"`
```
Oh, I thought the cluster send bypassed that somehow since I think I based this on your PR. I'll take a look at this again since I want to make sure the user doesn't get this

agnivade replied:
I didn't get, do you _not_ want to pass this to other nodes in the cluster?

I did that in my PR because that was a function, and you cannot marshal a function, which was the bug.

hmhealey replied:
I want it to pass these between nodes in the cluster, but I don't want them to be sent to the client since both of those use the current struct tags. I'll see if there's a way to easily handle those cases differently

agnivade replied:
Ah I see, in that case, you can just use `omitempty` and wipe the fields from the message once the hooks are executed. It will get omitted when json marshalled.

hmhealey commented on server/channels/app/websocket_broadcast_hooks.go:
```diff
@@ -0,0 +1,62 @@
+				isFollower = pUtils.Contains(followers, webConn.UserId)
+			}
+
+			if hasMention || isFollower {
```
I'll remove it. I think I was expecting these two code paths to behave more similarly

hmhealey commented:
Just to reflect some progress here, the only thing left to do on this is making sure cross-server communication works since it currently doesn't. I got a local cluster running now, so I just need to figure out how to make that work without leaking the broadcast hook info to the client.

hmhealey commented:
Alright, so to summarize the changes I've made since Agniva's review:

1. I split the one `BroadcastHook` implementation into two, one for followers and one for mentions. I also refactored and renamed some stuff around that, but that logic is mostly the same.
2. `WebsocketBroadcast.BroadcastHooks` and `.BroadcastHookArgs` are now serialized so that they're passed around the cluster. To make sure that those aren't sent to the client, those fields are now removed by the WebHub before it calculates the precomputed JSON, and they're now passed into `runBroadcastHooks` separately
3. I added some tests for the logic in that last set of changes. The order of these function calls matters and makes this logic a bit fragile, so hopefully the tests will prevent any potential future regressions.

streamer45 commented on server/public/model/websocket_message.go:
```diff
@@ -104,6 +104,17 @@ type WebsocketBroadcast struct {
+	// BroadcastHooks is a slice of hooks IDs used to process events before sending them on individual connections. The
+	// IDs should be understood by the WebSocket code.
+	//
+	// This field should never be sent to the client.
```
I see you took the other route. Not against it but we are making it "insecure by default" this way. If it helps keeping complexity low it's acceptable I'd say, but something to keep in mind.

hmhealey replied:
Yeah, I ended up going with this because it seems to work, and I've spent far longer on this than I'd expected to originally. As cool as the hook idea that I took from Agniva's other PR is, this kind of blew up in complexity, so I needed to find something that works to ship. In retrospect, I probably should've just gone with sending individual events to mentioned users to fix the original bug without adding all this.

As for making it more secure, I still think the best way to go about that would be to have two separate methods on the event for serializing it for server-server transmission and server-client transmission, but since this has already gotten so big, I didn't want to have to change that whole interface

streamer45 commented on server/public/model/websocket_message.go:
```diff
@@ -190,10 +208,38 @@ func (ev *WebSocketEvent) PrecomputeJSON() *WebSocketEvent {
+// AddWithCopy copies the map and writes to a copy of that,
+// and sets the map to the new event.
+func (ev *WebSocketEvent) AddWithCopy(key string, value any) {
+	ev.data = copyMap(ev.data)
```
This looks potentially racy.

hmhealey replied:
At the moment, it's not because we copy the event before making any changes using it, but I could definitely see that being a risk in the future.

@agnivade This we should change this now? I could also add a comment to which of these methods aren't thread-safe as a warning

hmhealey replied:
I ended up removing this and just doing a deep copy of the event before the hooks process it. That seems safer, and the cost of deep copying is so minimal when a hook wants to make changes that I don't think it will matter

streamer45 commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -0,0 +1,109 @@
+	return msg
+}
+
+func UseAddMentionsHook(message *model.WebSocketEvent, mentionedUsers model.StringArray) {
+	message.GetBroadcast().AddHook(broadcastAddMentions, map[string]any{
```
I see a couple of assumptions here and above, both message and broadcast not being `nil`. I guess that should be the case now but just want to make sure we are also considering the use of this in the future.

hmhealey replied:
`message` could be nil, but `broadcast` can't as long as `MakeNewWebSocketEvent` was used to create the event. I can see nil-checking this being good, but I don't think we typically do that for most parameters in the app, so I'd rather not add it here for a case that "shouldn't happen".

I mostly added these functions mostly to give a bit of type safety to this code, and there might be a way to do this by using some generic nonsense (like associating the `broadcastAddMentions` constant with a concrete arguments type) which would avoid needing a new helper function entirely, but that's potentially more scope creep on this PR

streamer45 commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -0,0 +1,109 @@
+func (h *addMentionsBroadcastHook) Process(msg *model.WebSocketEvent, webConn *platform.WebConn, args map[string]any) *model.WebSocketEvent {
+	mentions, ok := args["mentions"].(model.StringArray)
+	if !ok {
+		mlog.Warn("Invalid mentions value passed to addMentionsBroadcastHook", mlog.Any("mentions", args["mentions"]))
```
Is there a reason for not returning a proper error instead?

hmhealey replied:
Not particularly. I think I had a good reason before, but I can't think of it now. I'll change that

streamer45 commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -0,0 +1,109 @@
+
+const (
+	broadcastAddMentions  = "add_mentions"
+	broadcastAddFollowers = "addfollowers"
```
I'd expect this to be `add_followers`.

hmhealey replied:
Oh yeah, that should be. I'll fix that

streamer45 commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	if isFollower {
+		// Note that the client expects this field to be stringified
```
Unrelated but I'd love if we could plan something around this. We are still sending posts as JSON in JSON if I recall correctly.

hmhealey replied:
Yeah, I would as well. I'm making a list of things I'd want from a WebSocket API v2 that I've added it to

streamer45 commented on server/channels/app/platform/web_broadcast_hook.go:
```diff
@@ -0,0 +1,56 @@
+	for i, hookID := range hookIDs {
+		hook := h.broadcastHooks[hookID]
+		args := hookArgs[i]
+		if hook == nil {
```
Is this case worth logging? I'd expect it not to happen in practice.

hmhealey replied:
Yep, I'll add it just in case. Cases we expect not to happen seem like the best things to log if there's no way to check them at compile time

isacikgoz replied:
Maybe we can reduce the log level, it doesn't seem to be an actionable item by the administrator. Thoughts?

hmhealey replied:
Makes sense. I'll drop it to warning for now, but if you think it should go lower, I can do that

streamer45 commented on server/channels/app/platform/web_broadcast_hook.go:
```diff
@@ -0,0 +1,56 @@
+	// Shallowly copy the event and remove any precomputed JSON since one or more hooks wants to make changes to it
+	msg = msg.RemovePrecomputedJSON()
+
+	for i, hookID := range hookIDs {
```
Would it be feasible to merge the two loops into one?

hmhealey replied:
No, that was done intentionally. I wanted to make sure that the event is copied if and only if a hook wants to make changes, and since there's a lot of ways to change `WebSocketEvent` in a non-thread-safe way, I wanted this function to be responsible for copying the object when a change is to be made. That's also why `ShouldProcess` and `Process` are two separate methods

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -492,14 +495,19 @@ func (h *Hub) Start() {
 				}
+
+				// Remove the broadcast hook information before precomputing the JSON so that those aren't included in it
+				msg, broadcastHooks, broadcastHookArgs := msg.RemoveBroadcastHooks()
```
Looks like we'd be calling this and making a copy for each message, regardless of whether the hooks are set? If so it may have a performance impact given this is a particularly hot path.

hmhealey replied:
Great catch. I was specifically trying to be careful about making new copies in this code. I'll fix that

streamer45 reviewed (commented):
Nice work :+1: Really appreciate the healthy proportion between application code and tests.

Gave this a very quick first pass and left a few comments.

agnivade commented:
@isacikgoz - Since Claudio is out, I added you. This is a PR modifying the websocket code, which is very sensitive to performance. We need to give this a thorough review.

@hmhealey - This will need a load test to ensure there is no regression at high scale. Let me know if you are interested to run one. Otherwise, I can do it myself, but will take some time. 0/5 on either.

agnivade commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -0,0 +1,106 @@
+		return false, nil
+	}
+
+	mentions, ok := args["mentions"].(model.StringArray)
+	if !ok {
+		return false, fmt.Errorf("Invalid mentions value passed to addMentionsBroadcastHook: %v", args["mentions"])
+	}
```
This doesn't work cross-cluster because once the message is unmarshalled from a byte slice, it loses any type info. So mention parsing fails in a cluster scenario with:

```
error [2023-10-16 05:50:30.901 Z] runBroadcastHooks: Encountered error running ShouldProcess for broadcast hook caller="platform/web_broadcast_hook.go:37" error="Invalid mentions value passed to addMentionsBroad
castHook: [q1gtsishkfr388sa19o1iitspr]"
```

To fix this, we need to marshal and unmarshal again like:

```diff
diff --git a/server/channels/app/web_broadcast_hooks.go b/server/channels/app/web_broadcast_hooks.go
index 7c6a51c4ab..77810db878 100644
--- a/server/channels/app/web_broadcast_hooks.go
+++ b/server/channels/app/web_broadcast_hooks.go
@@ -4,6 +4,7 @@
 package app
 
 import (
+	"encoding/json"
 	"fmt"
 
 	"github.com/mattermost/mattermost/server/public/model"
@@ -30,24 +31,46 @@ func (h *addMentionsBroadcastHook) ShouldProcess(msg *model.WebSocketEvent, webC
 		return false, nil
 	}
 
-	mentions, ok := args["mentions"].(model.StringArray)
-	if !ok {
-		return false, fmt.Errorf("Invalid mentions value passed to addMentionsBroadcastHook: %v", args["mentions"])
+	if args["mentions"] == nil {
+		return false, nil
+	}
+
+	buf, err := json.Marshal(args["mentions"])
+	if err != nil {
+		return false, err
 	}
 
+	var mentionArray model.StringArray
+	if err = json.Unmarshal(buf, &mentionArray); err != nil {
+		return false, err
+	}
+
+	fmt.Println("PASSED---------", mentionArray)
+
 	// This hook will only modify the event if the current user is mentioned by the post
-	return pUtils.Contains[string](mentions, webConn.UserId), nil
+	return pUtils.Contains[string](mentionArray, webConn.UserId), nil
 }
 
 func (h *addMentionsBroadcastHook) Process(msg *model.WebSocketEvent, webConn *platform.WebConn, args map[string]any) (*model.WebSocketEvent, error) {
-	mentions, ok := args["mentions"].(model.StringArray)
-	if !ok {
-		return msg, fmt.Errorf("Invalid mentions value passed to addMentionsBroadcastHook: %v", args["mentions"])
+	if args["mentions"] == nil {
+		return nil, nil
+	}
+
+	buf, err := json.Marshal(args["mentions"])
+	if err != nil {
+		return nil, err
 	}
 
+	var mentionArray model.StringArray
+	if err = json.Unmarshal(buf, &mentionArray); err != nil {
+		return nil, err
+	}
+
+	fmt.Println("PASSED---------", mentionArray)
 	hasMention := false
-	if len(mentions) > 0 {
-		hasMention = pUtils.Contains(mentions, webConn.UserId)
+
+	if len(mentionArray) > 0 {
+		hasMention = pUtils.Contains(mentionArray, webConn.UserId)
 	}
 
 	if hasMention {

```

hmhealey replied:
Oh, that's kind of gross I tested locally, but that must've been before I checked `ok` since it would've just failed silently then.

I wish there was a better way to reinterpret the type of that so hopefully it won't add too much overhead to use JSON to reinterpret the type. I'll make that change, thanks

hmhealey replied:
I put this logic in a helper since I ended up repeating myself. It also has the nice benefit of letting us include skip the marshaling/unmarshaling step if the value is already the correct type which is always the case for single server setups.

agnivade reviewed (changes requested):
Gave it a first round of testing. We will dive deeper in the next round once the marshalling problem is fixed.

hmhealey commented:
Also, if you could run a load test for me once I've fixed that, I'd really appreciate it. It's been years since I've run one myself, so I'm definitely not set up to run them myself

agnivade commented:
Can do :+1:

agnivade commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -0,0 +1,134 @@
+	return msg, nil
+}
+
+func UseAddMentionsHook(message *model.WebSocketEvent, mentionedUsers model.StringArray) {
```
1/5, this and the other function does not need to be exposed.

agnivade reviewed (commented):
Gave it a deeper look this time. Couple of things to improve, but mostly looking good.

hmhealey commented on server/channels/app/platform/web_broadcast_hook.go:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	// Copy the event and remove any precomputed JSON since one or more hooks wants to make changes to it
+	msg = msg.DeepCopy().RemovePrecomputedJSON()
```
Yeah, that makes sense. I was a bit worried about changing the semantics of `RemovePrecomputedJSON`, but I forgot we were the ones that added it

hmhealey commented on server/channels/app/platform/web_broadcast_hook.go:
```diff
@@ -0,0 +1,67 @@
+	for i, hookID := range hookIDs {
+		hook := h.broadcastHooks[hookID]
+		args := hookArgs[i]
+		if hook == nil {
+			continue
+		}
```
That's a very good point that the current code assumes `Process` is only called when `ShouldProcess` is true even though that's not the case. I'll change it to track `hasChanges` on a per-hook basis.

We're also currently fixing the type of the hook arguments twice by going through the json encoder twice which I wasn't a huge fan of, and if we're tracking `hasChanges` on a per-hook basis, we could also do the same with the typed arguments as well

agnivade replied:
> We're also currently fixing the type of the hook arguments twice by going through the json encoder twice which I wasn't a huge fan of, and if we're tracking hasChanges on a per-hook basis, we could also do the same with the typed arguments as well

Yes indeed! We should do that only once.

hmhealey commented on server/channels/app/platform/web_hub.go:
```diff
@@ -70,6 +70,8 @@ type Hub struct {
 	explicitStop    bool
 	checkRegistered chan *webConnSessionMessage
 	checkConn       chan *webConnCheckMessage
+
```
It was a mental grouping thing of putting related fields together. I know the rest of the struct doesn't do that, but I think there might have been multiple fields related to this at some point.

I'll remove that line for consistency

hmhealey commented:
I ended up reworking how we ensure that hooks aren't racey, are mostly type-safe, and only copy the `WebSocketEvent` when necessary.

Previously, the `BroadcastHook` interface had separate `ShouldProcess` and `Process` methods so that `runBroadcastHooks` could handle copying the event and preventing racey behaviour. Now, `BroadcastHook` just has a single method which receives a `HookedWebSocketEvent`, a new type that wraps the `WebSocketEvent` to prevent racey behaviour and to ensure that the event is only copied when changes need to be made.

Having just a single method on the `BroadcastHook` also means that it doesn't need to duplicate the data of checking if it needs to process an event, and it also means that we only need to fix the types of the arguments once. It hopefully also just simplifies some stuff.

Let me know how this looks!

isacikgoz commented on server/channels/app/notification_test.go:
```diff
@@ -212,6 +217,309 @@ func TestSendNotifications(t *testing.T) {
 	})
 }
 
+func TestSendNotifications_MentionsFollowers(t *testing.T) {
```
I don't see much places we use underscores, In fact we could put this function as a `t.Run()` inside the `TestSendNotifications`. This is just a cosmetic concern, 0/5 so take it as a grain of salt.

hmhealey replied:
I split it out into a separate function because the setup for these are different. I suppose we could've nested it deeper, but that didn't feel great. Only 1/5 that I like keeping it in a separate function, but if there's a better naming convention for that, I could change it. I see there's also `TestSendNotificationsWithManyUsers` above which is similar

isacikgoz commented on server/channels/app/notification_test.go:
```diff
@@ -212,6 +217,309 @@ func TestSendNotifications(t *testing.T) {
+			ChannelId: th.BasicChannel.Id,
+			Message:   "@" + th.BasicUser.Username,
+		}
+		_, err = th.App.SendNotifications(th.Context, post, th.BasicTeam, th.BasicChannel, th.BasicUser, nil, false)
```
Uh, it seemed a bit weird to send a ws message to own user :/ Is it necessary?

hmhealey replied:
Oh that's a typo. These are supposed to be posting as `sender`, not as `th.BasicUser`. I'll fix those, and hopefully it doesn't cause the tests to mysteriously break

isacikgoz commented on server/channels/app/platform/web_hub.go:
```diff
@@ -492,14 +494,19 @@ func (h *Hub) Start() {
 					if webConn.ShouldSendEvent(msg) {
 						select {
-						case webConn.send <- msg:
+						case webConn.send <- h.runBroadcastHooks(msg, webConn, broadcastHooks, broadcastHookArgs):
```
Is it necessary to run the hooks while sending? Wondering whether hooks should run the hooks here or should we run those where it's been received?

hmhealey replied:
I think this is probably the best spot for it for now. I do agree that running it in the `WebConn` does feel right because we want these to run right before an event is sent to the client, but that also requires registering each broadcast hook with every `WebConn` which feels like we're passing that information around a lot. Because of that, I think it's fine for the `WebHub` to be running them. Only 1/5 though.

isacikgoz replied:
I see, thanks for the explanation.

isacikgoz reviewed (commented):
Overall this looks good to me, obviously it would be better to look at the load test results. I have a few questions below. Great work @hmhealey 👍

agnivade reviewed (approved):
Nicely done. Pending load tests.

isacikgoz reviewed (approved):
LGTM 👍

yasserfaraazkhan commented:
/e2e-tests

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/384100) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/384100)

yasserfaraazkhan reviewed (approved):
tested this by posting UserA's keywords from UserB's account. did not see the keywords of UserA showing up in the dev console of the UserB account.

agnivade commented:
@yasserfaraazkhan - Please test this in a HA spinwick as well. There are some edge cases there which we need to test and confirm. E2E tests should also run in HA.

agnivade commented:
I am running a loadtest for this now. Will post the results on Monday.

agnivade commented:
![websocket-buffer-size](https://github.com/mattermost/mattermost/assets/1774000/5cba84be-1e17-4240-bdec-748397ade74f)
![websocket-connections](https://github.com/mattermost/mattermost/assets/1774000/4dc662eb-4414-4911-9dfe-96b816fdb01c)
![p99-api-times](https://github.com/mattermost/mattermost/assets/1774000/140f324b-b227-42a5-bb6b-5cc12be79c41)
![avg-api-times](https://github.com/mattermost/mattermost/assets/1774000/2d2a074d-6267-4609-985e-4e5ec9e3be1b)
![p99-store-times](https://github.com/mattermost/mattermost/assets/1774000/1586e78b-c3bc-4e90-9cc8-d61a29fc2fbf)
![avg-store-times](https://github.com/mattermost/mattermost/assets/1774000/1136c96c-83bf-4966-8f61-2954681af24b)
![goroutines-in-use](https://github.com/mattermost/mattermost/assets/1774000/164051e8-0ad2-4125-a697-2ebe9ab8eb37)
![heap-in-use](https://github.com/mattermost/mattermost/assets/1774000/96b941ce-d731-4fba-bfc3-b24834ea43aa)
![cpu-utilization](https://github.com/mattermost/mattermost/assets/1774000/0dcf15c0-d19b-4721-89ac-a905809887e7)
![p50-createpost](https://github.com/mattermost/mattermost/assets/1774000/4b081c95-e0c1-4ebd-9fcd-36d1e0373271)
![p99-createpost](https://github.com/mattermost/mattermost/assets/1774000/579a6ecb-a71e-432a-8729-8ed0ce626dcf)

Difference looks to be within bounds. Probably needed 2 more runs per test to confirm, but there's nothing alarming.

hmhealey commented:
Awesome, thanks for running those for me, @agnivade! Does that mean it's good to merge now or are you wanting to run more tests to double check?

agnivade commented:
I've removed the label, so yes it's good to merge.

amyblais commented:
/cherry-pick release-8.1

mattermost-build commented:
Cherry pick is scheduled.

mattermost-build commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-#24641-upstream-release-8.1-1700250995
Switched to a new branch 'automated-cherry-pick-of-mattermost-#24641-upstream-release-8.1-1700250995'
Branch 'automated-cherry-pick-of-mattermost-#24641-upstream-release-8.1-1700250995' set up to track remote branch 'release-8.1' from 'upstream'.

+++ About to attempt cherry pick of PR #24641 with merge commit ef66f7beabb2429ef714720dee12d11847684204.

Auto-merging server/channels/app/notification.go
Auto-merging server/channels/app/notification_test.go
CONFLICT (content): Merge conflict in server/channels/app/notification_test.go
Auto-merging server/channels/app/platform/helper_test.go
Auto-merging server/channels/app/platform/service.go
Auto-merging server/channels/app/platform/web_hub.go
Auto-merging server/channels/app/platform/web_hub_test.go
Auto-merging server/channels/app/server.go
Auto-merging server/public/model/websocket_message.go
error: could not apply ef66f7beab... MM-54238 Add WebSocket broadcast hook and don't broadcast when other users were mentioned (#24641)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

+++ Conflicts detected:

UU server/channels/app/notification_test.go
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```
